File: web-ts/src/strings/NString.ts
/**
 * NString - String node type.
 * @JAVA_REF org.modelingvalue.nelumbo.strings.NString
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { TokenType } from '../syntax/TokenType';

const DELIM = '"';

let FUNCTOR: Functor | null = null;

export function setNStringFunctor(f: Functor): void {
  FUNCTOR = f;
}

export class NString extends Node {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    if (args.length === 1 && typeof args[0] === 'string') {
      // Constructor from parsed string - need to strip quotes
      const val = args[0] as string;
      super(functor, elements, strip(val) ?? val);
    } else {
      super(functor, elements, ...args);
    }
  }

  protected static fromData(data: unknown[], declaration?: Node): NString {
    const node = Object.create(NString.prototype) as NString;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Node | undefined })._declaration = declaration ?? node;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Node): Node {
    return NString.fromData(data, declaration ?? this.declaration());
  }

  static of(val: string): NString {
    if (FUNCTOR === null) {
      throw new Error('NString FUNCTOR not initialized');
    }
    return new NString(FUNCTOR, List(), val);
  }

  static strip(val: string | null): string | null {
    return val !== null && val.startsWith(DELIM)
      ? val.substring(1, val.length - 1)
      : null;
  }

  value(): string {
    return this.get(0) as string;
  }

  override toString(_previous?: TokenType[]): string {
    return DELIM + this.value() + DELIM;
  }
}

function strip(val: string): string | null {
  return NString.strip(val);
}

File: src/main/java/org/modelingvalue/nelumbo/strings/NString.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.strings;

import java.io.Serial;

import org.modelingvalue.nelumbo.ConstructionReason;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.NelumboFunctorField;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.NodeInfo;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.syntax.ParseContext;
import org.modelingvalue.nelumbo.syntax.ParseException;
import org.modelingvalue.nelumbo.syntax.TokenType;

public final class NString extends Node {
    @Serial
    private static final long serialVersionUID = 8360866611309554234L;

    private static final String DELIM = "\"";

    @NelumboFunctorField
    private static Functor FUNCTOR;

    @NelumboConstructor
    public NString(NodeInfo nodeInfo, Object... args) {
        super(nodeInfo, args);
    }

    public static NString of(String val) {
        return new NString(NodeInfo.of(FUNCTOR), val);
    }

    private static String strip(String val) {
        return val.substring(1, val.length() - 1);
    }

    @Override
    public String toString(TokenType[] previous) {
        return DELIM + getVal(0) + DELIM;
    }

    @Override
    public Node init(KnowledgeBase knowledgeBase, ParseContext ctx, ConstructionReason reason) throws ParseException {
        if (reason == ConstructionReason.parsing && get(0) instanceof String val && val.startsWith(DELIM)) {
            return set(0, strip(val));
        }
        return this;

    }

}
